src/upsample.jl

```diff
@@ -64,235 +64,166 @@ function ChainRulesCore.rrule(::typeof(upsample_nearest), x::AbstractArray, s::T
+# Outputs
+- `dx`: Downsampled version of `Δ`
+"""
+function ∇upsample_bilinear(Δ::AbstractArray{T,4}; size::NTuple{2,Integer}) where T
+    w, h, c, n = Base.size(Δ)
+    out_w, out_h = size
+    if (w,h) == (out_w, out_h)
+        return Δ
     end
-
-    return dx
+    dx = zero(similar(Δ, T, out_w, out_h, c, n))
+    return ∇upsample_bilinear_whcn!(dx, Δ)
 end
 
-# `n` upsample factor for which a downsample kernel will be determined.
-# Δ is given in case of necessity of gpu conversion 
-function get_downsamplekernel(Δ, n::Int)
-    step = 1//n
-    if n % 2 == 0
-        start = step//2
-        upward = collect(start:step:1//1)
-        kernel = [upward; reverse(upward)]
-    else
-        start = step
-        upward = collect(start:step:1//1)
-        kernel = [upward; reverse(upward[1:end-1])]
+function ∇upsample_bilinear_whcn!(dx::AbstractArray{T,4}, Δ::AbstractArray{T,4}) where T
+    size(dx)[3:4] == size(Δ)[3:4] || error("Number of input and output channels and batches must match. Got input $(size(input)) and output $(size(output))")
+    in_w, in_h, channels, batches = size(dx)
+
+    # treat batch and channel dimension as one for better parallelization granularity
+    channels *= batches
+    out_w, out_h, _, _ = size(Δ)
+    output_slice_size = out_h * out_w
+
+    width_scale  = T((in_w - 1) // (out_w - 1))
+    height_scale = T((in_h - 1) // (out_h - 1))
+
+    @inline idx(c, h, w) = c * in_h * in_w + h * in_w + w + 1
+
+    @inbounds Threads.@threads for c in 0:channels-1
+        for oh in 0:out_h-1
+            ih0, ih1, h0lambda, h1lambda = compute_source_index_and_lambda(height_scale, oh, in_h, out_h)
+            for ow in 0:out_w-1
+                iw0, iw1, w0lambda, w1lambda = compute_source_index_and_lambda(width_scale, ow, in_w, out_w)
+                output_offset = c * output_slice_size + oh * out_w + ow + 1
+                Δ_value = Δ[output_offset]
+                dx[idx(c, ih0, iw0)] += h0lambda * w0lambda * Δ_value # i00
+                dx[idx(c, ih0, iw1)] += h0lambda * w1lambda * Δ_value # i01
+                dx[idx(c, ih1, iw0)] += h1lambda * w0lambda * Δ_value # i10
+                dx[idx(c, ih1, iw1)] += h1lambda * w1lambda * Δ_value # i11
+            end
+        end
     end
-    # TODO there must be a more convenient way to send to gpu 
-    kernel = convert(typeof(Δ), reshape(kernel, length(kernel), 1, 1, 1))
-    kernel = dropdims(kernel, dims=(2,3,4))
-    return kernel
+    return dx
 end
 
-function ChainRulesCore.rrule(::typeof(upsample_bilinear), x, k)
-    Ω = upsample_bilinear(x, k)
+function ChainRulesCore.rrule(::typeof(upsample_bilinear), x; size)
+    Ω = upsample_bilinear(x; size=size)
     function upsample_bilinear_pullback(Δ)
-        (NO_FIELDS, ∇upsample_bilinear(Δ, k), DoesNotExist())
+        (NO_FIELDS, ∇upsample_bilinear(Δ; size=(Base.size(x,1),Base.size(x,2))), DoesNotExist())
```

Comment:
```suggestion
        (NO_FIELDS, ∇upsample_bilinear(Δ; size=(Base.size(x,1),Base.size(x,2))))
```

Reply:
Is it this that's meant to be committed?

Reply:
Good catch. Should we be disturbed that CI didn't catch this?

Reply:
If this is needed, to be correct, it should be positional no?

Reply:
this is a consequence of the method split in last commit

Reply:
The api should reflect that then

Reply:
I don't understand

Reply:
@DhairyaLGandhi can you be more explicit?

Reply:
If the methods were split, then there should be corresponding functions that should be able to completely specify the expected outcome with those new split methods.

Reply:
Is it concerning that CI didn't complain? 

There is a test which runs this gradient. Maybe charinrules etc. just takes the first N elements of the tuple, without checking its length, and so just ignores any trailing elements.

Seems far removed from API concerns, not sure what you are thinking. Nor why you suggest it should be positional? Of course things could be designed other ways, but at present the size is always a keyword.

Reply:
So you could have a method which take the dimensions as positional in this case.

Reply:
Yea, other designs are possible.

Reply:
It may not have been caught since we can't exactly guarantee the number of partials any such call might have. But yes, we should do the correct thing else we'll get the incorrect values, so no question on the change.

The split methods may be an artifact of refactoring, but the downstream functions should be consistent and if they need the size given explicitly in some cases, then they need to accept them, and the need aspect is determined by it being positional.

Reply:
> The split methods may be an artifact of refactoring, but the downstream functions should be consistent and if they need the size given explicitly in some cases, then they need to accept them, and the need aspect is determined by it being positional.

This is pretty hard to parse. Can you write a code sample showing us what you have in mind?

By "split" Carlo just means that there was a change from one function body with if-statements, to using multiple dispatch. He doesn't mean "broken" or something. Obviously things must work, if you think there are problems with "downstream functions" not getting what they need, then I think you ought to say exactly which function you mean. But revisiting the whole design is well beyond the scope of this one-line correction.